source/extensions/common/tap/admin.h

```diff
@@ -52,11 +63,117 @@ class AdminHandler : public Singleton::Instance,
   void unregisterConfig(ExtensionConfig& config);
 
   // Extensions::Common::Tap::Sink
-  PerTapSinkHandlePtr createPerTapSinkHandle(uint64_t) override {
-    return std::make_unique<AdminPerTapSinkHandle>(*this);
-  }
+  PerTapSinkHandlePtr
+  createPerTapSinkHandle(uint64_t,
+                         envoy::config::tap::v3::OutputSink::OutputSinkTypeCase type) override;
 
 private:
+  /**
+   * TraceBuffer internally buffers TraceWrappers in a vector. The size of the
+   * internal buffer is determined on construction by max_buf_size. TraceBuffer will continue to
+   * buffer traces via calls to bufferTrace until there is no remaining room in the buffer, or
+   * traceList is called, transfering ownership of the buffered data.
+   * Note that TraceBuffer is not threadsafe by itself - accesses to TraceBuffer need to be
+   * serialized. Serialization is currently done by the main thread dispatcher.
+   */
+  class TraceBuffer {
+    using TraceWrapper = envoy::data::tap::v3::TraceWrapper;
+    const size_t max_buf_size_; // Number of traces to buffer
+    absl::optional<std::vector<TraceWrapper>> buffer_;
```

Comment:
move variable declarations to a `private:` section at bottom of class decl per Envoy conventions